src/rpc/server.cpp

```diff
@@ -550,6 +550,51 @@ std::string experimentalDisabledHelpMsg(const std::string& rpc, const std::vecto
         + config;
 }
 
+std::string asOfHeightMessage(bool hasMinconf) {
+    std::string minconfInteraction = hasMinconf
+        ? "                    `minconf` must be at least 1 when `asOfHeight` is provided.\n"
+        : "";
+    return
+        ". asOfHeight       (numeric, optional, default “*”) Execute the query as if it\n"
+        "                    were run when the blockchain was at the height specified by\n"
+        "                    this argument. The default is to use the entire blockchain\n"
+        "                    that the node is aware of. A “future” height will fall back\n"
+        "                    to the current height. Any explicit value will cause the\n"
+        "                    mempool to be ignored, meaning no unconfirmed tx will be\n"
+        "                    considered.\n"
+        + minconfInteraction;
+}
+
+std::optional<int> parseAsOfHeight(const UniValue& params, int index) {
+    std::optional<int> asOfHeight;
+    if (params.size() > index && !(params[index].isStr() && params[index].get_str() == "*")) {
+        auto requestedHeight = params[index].get_int();
+        if (requestedHeight < 0) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Can not perform the query as of a negative block height");
+        } else if (requestedHeight == 0) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Can not perform the query as of the genesis block");
+        } else {
+            asOfHeight = params[index].get_int();
+        }
+    }
+    return asOfHeight;
+}
```

Comment:
Other height parameters accept a negative argument, for example [getblock](https://zcash.github.io/rpc/getblock.html):
```
1. "hash|height"          (string, required) The block hash or height. Height can be negative where -1 is the last known valid block
```
This would allow the default to be "-1" rather than defining a special "*" value.

An argument against this approach is that APIs like `getblock` (for verbosity >= 1) return the resolved height in the result, which makes them more resistant to race conditions, whereas that can't be done for methods that don't return a JSON object. I'm fine with leaving it as-is if you think it's a bad idea.

```suggestion
int parseAsOfHeight(const UniValue& params, int index, int currentHeight) {
    int requestedHeight = currentHeight;
    if (params.size() > index) {
        requestedHeight = parseHeightArg(params[index].get_str(), currentHeight);
    }
    if (requestedHeight == 0) {
        throw JSONRPCError(RPC_INVALID_PARAMETER, "Can not perform the query as of the genesis block");
    }
    return requestedHeight;
}
```

Reply:
I like this idea. However, it needs a corresponding change in wallet.cpp to ensure we don’t resolve to a negative height, and I think one or two other things. So I’m not going to merge this suggestion as-is.

Reply:
So, for `-1` to be the default, we still need to handle that value specially to avoid it ignoring the mempool. I’m thinking we can resolve the height in rpcwallet (similarly to how `getblock` does), then if `asOfHeight == currentHeight`, use `std::nullopt`.

But, as you mentioned, doing anything that checks the active height at this point can cause a race condition, and we don’t have the resolved block (height) in the response to make it more resilient. In which case, we could _not_ allow negatives in general, but still allow `-1` as “current height” to be consistent with the other height params.

I’m leaning toward the latter at the moment.